content/blog/fp-js-higher-order-functions.md

```diff
@@ -0,0 +1,577 @@
+
+For example, say we have a pretty dashboard, full of beautiful graphs that display meaningful data. Let's make it a line chart, that displays multiple lines, and we want to have the ability to toggle certain lines on and off.
+
+The data could look something like this:
+
+```js
+const veryImportantData = [
+  {name: 'Impressions', visible: true, data: [...]},
+  {name: 'Clicks', visible: false, data: [...]},
+  {name: 'PageViews', visible: true, data: [...]}
+]
+```
+
+We have already built a function using our newfound knowledge of `map` to toggle the `visible` state on and off for a line, but now we need to grab only the lines that have `visible: true`, and send them into our chart for presentation. This is where `filter` comes in handy!
+
+```js
+function isLineVisible(line) {
+  return line.visible
+}
+
+const dataToPresent = veryImportantData.filter(isLineVisible)
+```
+
+Or, we could do it inline with an anonymous function on one line!
+
+```js
+const dataToPresent = veryImportantData.filter(line => line.visible)
+```
+
+Which gives us a new array with only the lines that had `visible: true`. The rest have been filtered out. Pretty great, right?
+
+Compare this to an approach using `for` loops:
+
+```js
+const dataToPresent = []
+
+for (let i = 0; i < veryImportantData.length; i++) {
+  if (veryImportantData[i].visible) {
+    dataToPresent.push(veryImportantData[i])
+  }
+}
+```
+
+Once again, the higher-order function version is easier to grasp, and we don't need to care about how the loop is working, or how the array is created. It's all abstracted away from us, focusing on the important part - filtering logic.
+
+### Array.prototype.reduce
+
+`Array.prototype.reduce`, while very useful, is a bit harder to grasp than `Array.prototype.map` and `Array.prototype.filter`, it is very versatile, and can be used to build new objects, numbers, strings, or arrays.
+
+Just as the previous functions (`map`, `filter`), `reduce` takes a callback function - called a reducer function - as the first argument. However, this callback function has a slight difference from the functions passed to `map` and `filter`. It takes an `accumulator` as the first argument, and the element it's currently operating (`currentValue`) on as the second. The `reduce` function also has a second argument, `initialValue`, which can be sent in to set the `accumulator` to have a value the first iteration. `reduce` is often referred to as `fold` withing other functional programming languages.
+
+Let's look at an example - our shopping cart. Say you want to present the total price of the shopping cart. This is a handy use for reduce!
+
+```js
+const addProductPriceToTotal = (total, product) => total + product.totalPrice
+
+const toalPrice = products
+  .map(addTotalPriceToProduct)
+  .reduce(addProductPriceToTotal, 0)
```

Comment:
Maybe first show `reduce` without `map.reduce`. I like the combination of both but I think it would be good to display just the use of `reduce` and then the combination after